lib/irslackd.js

```diff
@@ -726,6 +745,22 @@ class Irslackd {
       this.ircd.write(ircUser.socket, ircUser.ircNick, 'PART', [ ircChan ]);
     }
   }
+  updatePresenceSubscriptions(ircUser) {
+    if (!ircUser.presenceEnabled()) {
+      return;
+    }
+    let userIds = new Set();
+    ircUser.channelNicks.forEach((nicks, channel) => {
+      nicks.forEach((_, nick) => {
+        let userId = ircUser.ircToSlack.get(nick);
+        if (userId) {
+          userIds.add(userId);
+        }
+      });
+    });
+    console.log('Subscribing to presence of', userIds.size, 'users');
+    ircUser.slackRtm.subscribePresence(Array.from(userIds));
```

Comment:
I think we should also call `subscribePresence` when someone new joins a channel the user is in. Either that or possibly there is a 'new user' event that we can subscribe to. Not sure if that event exists.

Looks great otherwise. Will merge with unit tests. Thanks again!

Reply:
Would the [team_join](https://api.slack.com/events/team_join) make sense for this?

Otherwise, I can call `updatePresenceSubscriptions` at the end of `onSlackMemberJoinedChannel` (this should be OK since most people don't constantly join and leave channels on Slack).

Reply:
I guess one problem with `team_join` is that the new member may not be in a channel the user is in and thus will not get a subscription.  This is OK, but if the new member later joins a channel the user is in, it will still not get a subscription.  Therefore, it is probably best to do the call at the end of `onSlackMemberJoinedChannel`. 

Let me know what you think.

Reply:
> I guess one problem with team_join is that the new member may not be in a channel the user is in and thus will not get a subscription.

Is that documented somewhere? Reading over the docs it seems like you are free to subscribe to any user in the workspace. I ask because doing this in `team_join` would be cleaner and more efficient.

Reply:
This is an implementation issue on our side.  The `updatePresenceSubscriptions` function only adds users that are in `channelNicks` (ie. nicks that were already in the room when the IRC client joins or nicks that have joined the room since the IRC client has joined).

It is possible that on `team_join` that the new member is not in any channel that the IRC client is in, and thus would never be added to `channelNicks`.

Practically speaking, this is probably not true since I believe everyone has to be in `#general` or the equivalent and thus a `onSlackMemberJoinedChannel` would be triggered.  Thus adding `updatePresenceSubscriptions` to the end of `onSlackMemberJoinedChannel` would take care of this.

I have been running the server with modification to `onSlackMemberJoinedChannel` for a few days and it seems fine (a few disconnects here and there, but presence monitoring works).

That said, I agree that using `team_join` would be more efficient.  I'm still open to either approach.

Reply:
Ah, I see what you mean. Instead of using `channelNicks` can we use the keys of `slackToIrc` and pick out anything that begins with a 'U'? If so I'd put this behind a method `IrcUser::getAllSlackUserIds` or something so that we can easily change the implementation down the line if desired. (In the future it may be desirable to split `slackToIrc` and `ircToSlack` maps into separate maps per object type (channel, user, bot, team, etc)). Then the `team_join` handler can simply invoke `mapIrcToSlack` and that should keep things updated. What do you think?